### PR TITLE
Client Side Image Optimization

### PR DESCRIPTION
**Pull request by xfalcox**
Integrates [mozJPEG](https://github.com/mozilla/mozjpeg) and [Resize](https://github.com/PistonDevelopers/resize) using WebAssembly to optimize user uploads in the composer on the client-side.

NPM libraries are sourced from our [Squoosh fork](https://github.com/discourse/squoosh/tree/discourse), which was needed because we have an older asset pipeline.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/services/media-optimization-worker.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
+import Service from "@ember/service";
+import { getOwner } from "@ember/application";
+import { Promise } from "rsvp";
+import { fileToImageData } from "discourse/lib/media-optimization-utils";
+import { getAbsoluteURL, getURLWithCDN } from "discourse-common/lib/get-url";
+
+export default class MediaOptimizationWorkerService extends Service {
+  appEvents = getOwner(this).lookup("service:app-events");
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/services/media-optimization-worker.js:8]: Can we not use the inject method for this?

File: app/assets/javascripts/discourse/app/services/media-optimization-worker.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  worker = null;
+  debugger;
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/services/media-optimization-worker.js:10]: Should probably remove this :)

File: app/assets/javascripts/discourse/app/services/media-optimization-worker.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    this.worker = new Worker(this.workerUrl); // TODO come up with a workaround for FF that lacks type: module support
+  }
+
+  stopWorker() {
+    this.worker.terminate();
+    this.worker = null;
+  }
+
+  ensureAvailiableWorker() {
+    if (this.worker === null) {
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/services/media-optimization-worker.js:25]:
```suggestion
    if (!this.worker) {
```

File: app/assets/javascripts/discourse/app/services/media-optimization-worker.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    }
+    if (
+      file.size <
+      this.siteSettings
+        .composer_media_optimization_image_kilobytes_optimization_threshold
+    ) {
+      return data;
+    }
+    this.ensureAvailiableWorker();
+    let p = new Promise(async (resolve) => {
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/services/media-optimization-worker.js:52]: Trivial but you should be able to `return new Promise..` instead of assigning it to a variable and returning that.

File: config/locales/server.en.yml [hunk 5]
```diff
@@ -1826,6 +1826,12 @@ en:
 
     strip_image_metadata: "Strip image metadata."
 
+    composer_media_optimization_image_enabled: "Enables client-side media optimization on uploaded image files."
```
Comment (eviltrout) [inline config/locales/server.en.yml:1829]:
```suggestion
    composer_media_optimization_image_enabled: "Enables client-side media optimization of uploaded image files."
```

File: config/locales/server.en.yml [hunk 6]
```diff
@@ -1826,6 +1826,12 @@ en:
...
+    composer_media_optimization_image_kilobytes_optimization_threshold: "Minimum image file size to trigger client-side optimization"
+    composer_media_optimization_image_resize_dimensions_threshold: "Minimum image width to trigger client-size resize"
```
Comment (eviltrout) [inline config/locales/server.en.yml:1831]:
```suggestion
    composer_media_optimization_image_resize_dimensions_threshold: "Minimum image width to trigger client-side resize"
```

File: config/locales/server.en.yml [hunk 7]
```diff
@@ -1826,6 +1826,12 @@ en:
...
+    composer_media_optimization_image_resize_width_target: "Images with width larger than `composer_media_optimization_image_dimensions_resize_threshold` will be resized to this width. Must be >= than `composer_media_optimization_image_dimensions_resize_threshold`."
```
Comment (eviltrout) [inline config/locales/server.en.yml:1832]:
```suggestion
    composer_media_optimization_image_resize_width_target: "Images with widths larger than `composer_media_optimization_image_dimensions_resize_threshold` will be resized to this width. Must be >= than `composer_media_optimization_image_dimensions_resize_threshold`."
```

File: public/javascripts/media-optimization-worker.js [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    };
+  }
+
+  return {
+    width: target_width,
+    height: Math.round((input_height / input_width) * target_width),
+  };
+}
+
+function logIfDebug(message, debugMode) {
```
Comment (eviltrout) [inline public/javascripts/media-optimization-worker.js:28]: Minor suggestion: Rather than passing `debugMode` in each time you call this, you could assign a local variable below like this:

```
  let log = settings.debug_mode ? console.log : () => {};
```

Then you can do `log('whatever')` after that and not worry about it.

**Comment by eviltrout** [comment]
Overall this is looking great. Very strong JS skills!

File: app/assets/javascripts/discourse/app/services/media-optimization-worker.js (hunk 1 above)
Comment (xfalcox) [reply app/assets/javascripts/discourse/app/services/media-optimization-worker.js:8]: There are new methods to use the service in newer versions of EmberJS, but this was the one that worked with a class in 3.15 in my tests. Also the one that is documented at https://guides.emberjs.com/v3.15.0/services/

File: public/javascripts/media-optimization-worker.js (hunk 8 above)
Comment (xfalcox) [reply public/javascripts/media-optimization-worker.js:28]: That was the first thing I tried, but the global variable scope inside a dedicated web worker was being weird. Let me try again.

File: app/assets/javascripts/discourse/app/services/media-optimization-worker.js (hunk 1 above)
Comment (eviltrout) [reply app/assets/javascripts/discourse/app/services/media-optimization-worker.js:8]: To be clear I'm talking about the method we often use in our codebase already: 

https://github.com/discourse/discourse/blob/master/app/assets/javascripts/admin/addon/components/themes-list.js#L8

That doesn't work in your class?

Comment (xfalcox) [reply app/assets/javascripts/discourse/app/services/media-optimization-worker.js:8]: I tried that and it didn't work. Maybe because I'm using native class versus the old Ember `Component.extend({` method?

Comment (eviltrout) [reply app/assets/javascripts/discourse/app/services/media-optimization-worker.js:8]: Weird! If you tried that and it didn't work we can leave for now. Eventually we'll get it working.

**Review by eviltrout — APPROVED** [review]

File: app/assets/javascripts/discourse/app/services/media-optimization-worker.js (hunk 1 above)
Comment (xfalcox) [reply app/assets/javascripts/discourse/app/services/media-optimization-worker.js:8]: I can clean it up to the newer idiom when we updated Ember!
